client/src/main/java/io/pravega/client/stream/impl/ReaderGroupStateManager.java

```diff
@@ -151,31 +147,40 @@ void close() {
     }
 
     /**
-     * Handles a segment being completed by calling the controller to gather all successors to the completed segment.
+     * Handles a segment being completed by calling the controller to gather all successors to the
+     * completed segment. To ensure consistent checkpoints, a segment cannot be released while a
+     * checkpoint for the reader is pending, so it may or may not succeed.
+     * 
+     * @return true if the completed segment was released successfully.
      */
-    void handleEndOfSegment(Segment segmentCompleted, boolean fetchSuccesors) throws ReaderNotInReaderGroupException {
+    boolean handleEndOfSegment(Segment segmentCompleted) throws ReaderNotInReaderGroupException {
         final Map<Segment, List<Long>> segmentToPredecessor;
-        if (fetchSuccesors) {
+        if (sync.getState().getEndSegments().containsKey(segmentCompleted)) {
+            segmentToPredecessor = Collections.emptyMap();
+        } else {
             val successors = getAndHandleExceptions(controller.getSuccessors(segmentCompleted), RuntimeException::new);
             segmentToPredecessor = successors.getSegmentToPredecessor();
-        } else {
-            segmentToPredecessor = Collections.emptyMap();
         }
 
         AtomicBoolean reinitRequired = new AtomicBoolean(false);
-        sync.updateState((state, updates) -> {
+        boolean result = sync.updateState((state, updates) -> {
             if (!state.isReaderOnline(readerId)) {
                 reinitRequired.set(true);
             } else {
                 log.debug("Marking segment {} as completed in reader group. CurrentState is: {}", segmentCompleted, state);
                 reinitRequired.set(false);
-                updates.add(new SegmentCompleted(readerId, segmentCompleted, segmentToPredecessor));
+                if (state.getCheckpointForReader(readerId) == null) {
```

Comment:
I'm not sure I understand this predicate. Is it checking whether there is an outstanding checkpoint for `readerId`? But, aren't we supposed to release completed segments (we are in the `handleEndOfSegment` method of the reader groups state manager) strictly after completing a checkpoint?

Reply:
This is guarding against the possible race condition of a second checkpoint being created at the last moment.